In [1]:
import pandas as pd

#1. Read the main PPI file (normalized_consensus_scores.csv)
#If it's comma-separated, use sep=",", otherwise sep="\t".
scores = pd.read_csv("./results_11032025/Jess_PPI_21032025/4_normalized_consensus_scores_Maxime_corr.csv", sep= ",", header=0)
scores.head()

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db  
0       1       1  
1       1       1  
2       1       1  
3       1       1  
4       1       1

In [2]:
#2. Read the ORF mapping file (tab-separated)
orf_map = pd.read_csv("./results_11032025/Jess_PPI_21032025/orf_symbol_ensg.csv", sep= "\t", header=0)
orf_map.head()

orf_id symbol     ensembl_gene_id
0       1  CALM2  ENSG00000143933.16
1       1  CALM1  ENSG00000198668.10
2       1  CALM3  ENSG00000160014.16
3       2    GBA  ENSG00000177628.15
4       3   BET1  ENSG00000105829.11

In [3]:
#3. Merge on db_orf_id
scores_db = scores.merge(
    orf_map,
    left_on="db_orf_id",   #Column in scores
    right_on="orf_id",     #Column in orf_map
    how="left"             #Keep all rows from 'scores'
)

scores_db.head()

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db  orf_id symbol     ensembl_gene_id  
0       1       1   52920  LITAF  ENSG00000189067.12  
1       1       1    2713   MLH1  ENSG00000076242.14  
2       1       1   10122  EXOC8   ENSG00000116903.7  
3       1       1    2713   MLH1  ENSG00000076242.14  
4       1       1    2713   MLH1  ENSG00000076242.14

In [4]:
#4. Rename columns from orf_map to indicate DB
scores_db.rename(
    columns={
        "symbol": "db_symbol",
        "ensembl_gene_id": "db_ensembl_gene_id"
    },
    inplace=True
)

scores_db.head() #double check 

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db  orf_id db_symbol  db_ensembl_gene_id  
0       1       1   52920     LITAF  ENSG00000189067.12  
1       1       1    2713      MLH1  ENSG00000076242.14  
2       1       1   10122     EXOC8   ENSG00000116903.7  
3       1       1    2713      MLH1  ENSG00000076242.14  
4       1       1    2713      MLH1  ENSG00000076242.14

In [5]:
#5. Drop the redundant 'orf_id' column
scores_db.drop("orf_id", axis=1, inplace=True)
scores_db.head()

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db db_symbol  db_ensembl_gene_id  
0       1       1     LITAF  ENSG00000189067.12  
1       1       1      MLH1  ENSG00000076242.14  
2       1       1     EXOC8   ENSG00000116903.7  
3       1       1      MLH1  ENSG00000076242.14  
4       1       1      MLH1  ENSG00000076242.14

In [6]:
#6. Merge again for ad_orf_id
scores_db_ad = scores_db.merge(
    orf_map,
    left_on="ad_orf_id",
    right_on="orf_id",
    how="left"
)

#7. Rename columns to indicate AD
scores_db_ad.rename(
    columns={
        "symbol": "ad_symbol",
        "ensembl_gene_id": "ad_ensembl_gene_id"
    },
    inplace=True
)

scores_db_ad.head()

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db db_symbol  db_ensembl_gene_id  orf_id ad_symbol  \
0       1       1     LITAF  ENSG00000189067.12    1235    HPCAL4   
1       1       1      MLH1  ENSG00000076242.14   11922      MYOG   
2       1       1     EXOC8   ENSG00000116903.7    7880  KIAA1217   
3       1       1      MLH1  ENSG00000076242.14    4351      NME4   
4       1       1      MLH1  ENSG00000076242.14   11671      FAN1   

   ad_ensembl_gene_id  
0  ENSG00000116983.12  
1   ENSG00000122180.4  
2  ENSG00000120549.17  
3  ENSG00000103202.12  
4   ENSG00000198690.9

In [7]:
#8. Drop the redundant 'orf_id' column again
scores_db_ad.drop("orf_id", axis=1, inplace=True)
scores_db_ad.head()

ad_orf_id  db_orf_id  db_mut_id retest_batch  normalized_score  assay_id  \
0       1235      52920       2676   VUSAPWT2B1             0.125         2   
1      11922       2713        953   VUSAPWT1B1             0.000         1   
2       7880      10122          0   VUSAPWT1B2             0.500         1   
3       4351       2713        869   VUSAPWT6B1             0.000         6   
4      11671       2713        455   VUSAPWT2B1             0.000         2   

   seq_ad  seq_db db_symbol  db_ensembl_gene_id ad_symbol  ad_ensembl_gene_id  
0       1       1     LITAF  ENSG00000189067.12    HPCAL4  ENSG00000116983.12  
1       1       1      MLH1  ENSG00000076242.14      MYOG   ENSG00000122180.4  
2       1       1     EXOC8   ENSG00000116903.7  KIAA1217  ENSG00000120549.17  
3       1       1      MLH1  ENSG00000076242.14      NME4  ENSG00000103202.12  
4       1       1      MLH1  ENSG00000076242.14      FAN1   ENSG00000198690.9

In [8]:
#9. Reorder columns for convenience: 

#Define desired column order in a list
new_order = [
    #DB columns
    'db_orf_id', 'db_symbol', 'db_ensembl_gene_id',
    'db_mut_id',
    
    #AD columns
    'ad_orf_id', 'ad_symbol', 'ad_ensembl_gene_id',
    
    #Rest 
    'retest_batch', 'normalized_score', 'assay_id',
    'seq_ad', 'seq_db'
]

#Reassign the df to the reordered version
scores_db_ad = scores_db_ad[new_order]
scores_db_ad.head()

db_orf_id db_symbol  db_ensembl_gene_id  db_mut_id  ad_orf_id ad_symbol  \
0      52920     LITAF  ENSG00000189067.12       2676       1235    HPCAL4   
1       2713      MLH1  ENSG00000076242.14        953      11922      MYOG   
2      10122     EXOC8   ENSG00000116903.7          0       7880  KIAA1217   
3       2713      MLH1  ENSG00000076242.14        869       4351      NME4   
4       2713      MLH1  ENSG00000076242.14        455      11671      FAN1   

   ad_ensembl_gene_id retest_batch  normalized_score  assay_id  seq_ad  seq_db  
0  ENSG00000116983.12   VUSAPWT2B1             0.125         2       1       1  
1   ENSG00000122180.4   VUSAPWT1B1             0.000         1       1       1  
2  ENSG00000120549.17   VUSAPWT1B2             0.500         1       1       1  
3  ENSG00000103202.12   VUSAPWT6B1             0.000         6       1       1  
4   ENSG00000198690.9   VUSAPWT2B1             0.000         2       1       1

In [9]:
#.10. Check if any db or ad_orf_id remained unmapped

#Find unmapped db_orf_ids => if db_symbol is NaN, that means the db_orf_id did not find a match in your ORF mapping file. 
unmapped_db = scores_db_ad.loc[scores_db_ad['db_symbol'].isna(), 'db_orf_id'].unique()

if len(unmapped_db) > 0:
    print("Unmapped DB orf_ids:", unmapped_db)
else:
    print("No unmapped DB orf_ids")

#11. Find unmapped ad_orf_ids
unmapped_ad = scores_db_ad.loc[scores_db_ad['ad_symbol'].isna(), 'ad_orf_id'].unique()

if len(unmapped_ad) > 0:
    print("Unmapped AD orf_ids:", unmapped_ad)
else:
    print("No unmapped AD orf_ids")

#12. Print out all rows where *either* db or ad is unmapped
unmapped_rows = scores_db_ad[
    scores_db_ad['db_symbol'].isna() | scores_db_ad['ad_symbol'].isna()
]

print("Rows with unmapped ORF IDs:\n", unmapped_rows)

No unmapped DB orf_ids
No unmapped AD orf_ids
Rows with unmapped ORF IDs:
 Empty DataFrame
Columns: [db_orf_id, db_symbol, db_ensembl_gene_id, db_mut_id, ad_orf_id, ad_symbol, ad_ensembl_gene_id, retest_batch, normalized_score, assay_id, seq_ad, seq_db]
Index: []


In [10]:
scores_db_ad.to_excel("./results_11032025/Jess_PPI_21032025/PPI_1percent_list_13042025.xlsx", sheet_name="MergedData", index=False)

In [11]:
#Get unique db_symbol genes
unique_db_symbols = scores_db_ad["db_symbol"].unique()
print("Unique db_symbol genes:")
print(unique_db_symbols)

Unique db_symbol genes:
['LITAF' 'MLH1' 'EXOC8' 'SUOX' 'PLA2G6' 'GRN' 'PRPF31' 'CDKN1A' 'ZMYND10'
 'PKP2' 'POLR1C' 'TPM1' 'TH' 'AGXT' 'UBQLN2' 'ACSF3' 'PLP1' 'PRKAR1A'
 'APOD' 'OPTN' 'KRT6A' 'MIP' 'TTPA' 'EIF2B4' 'PMP22' 'STXBP1' 'PNKP'
 'KRT2' 'AC012254.2' 'IER3IP1' 'ANKRD1' 'HMBS' 'NMNAT1' 'CLDN19' 'CTNNA3'
 'CCBE1' 'KRT5' 'GOSR2' 'SMAD3' 'SFTPC' 'KRT86' 'BAG3' 'EMD' 'SYNGR1'
 'CHN1' 'ESR2' 'ALAS2' 'TMEM43' 'GFAP' 'GCK' 'THAP1' 'ADIPOQ' 'KRT8'
 'RAD51D' 'GLYCTK' 'BLK' 'TPM3' 'KCNJ2' 'STAR' 'NFU1' 'HPRT1' 'FAM161A'
 'CCDC103' 'FAM187A' 'DMC1' 'PLOD3' 'ALDOA' 'SFTPA2' 'TRIM32' 'PRKRA'
 'DOLK' 'MID1' 'EFHC1' 'PRPF3' 'TTR' 'NAA10' 'COL2A1' 'KRT4' 'BFSP2'
 'DUSP23' 'COQ8A' 'KCTD7' 'DCX' 'OAS1' 'PYGL' 'GYPA' 'AMPD2' 'TIMM8A'
 'ATPAF2' 'FA2H' 'CLN6' 'PITX2' 'MVK' 'SH3BP2' 'PITX1' 'GNAI3' 'AP2S1'
 'RXRG' 'AHCY' 'DIABLO' 'VHL' 'PUF60' 'FADD' 'ITGB2' 'PHKG2' 'NCF2' 'ASNS'
 'LHFPL5' 'EFEMP1' 'IMPDH1' 'CDK4' 'PTS' 'APOA1' 'EIF2B1' 'IL36RN' 'GCM2'
 'PLN' 'SMARCB1' 'SLC30A2' 'TFG' 'KCTD1' 'CDC73' 

In [12]:
len(unique_db_symbols)

247

In [13]:
#List interacting partners per db_symbol:
#A straightforward approach is to use groupby on db_symbol and collect the unique ad_symbol values:

partners_per_gene = (
    scores_db_ad
    .groupby("db_symbol")["ad_symbol"]
    .unique()  #or .apply(list) if we want them as a list instead of a NumPy array
)

#'partners_per_gene' is a Series whose index is 'db_symbol' and 
#whose values are arrays of the corresponding 'ad_symbol' partners.

In [14]:
#Now we can iterate over partners_per_gene to print each db_symbol followed by its list of ad_symbol partners:

for db_symbol, ad_symbols in partners_per_gene.items():
    print(f"db_symbol: {db_symbol}")
    print(f"Interacting partners (ad_symbol): {list(ad_symbols)}\n")

db_symbol: AC012254.2
Interacting partners (ad_symbol): ['GPR42', 'TMEM237', 'PIGP', 'TMEM86B', 'HERPUD2', 'TMEM80', 'UBQLN1', 'TM4SF18']

db_symbol: ACSF3
Interacting partners (ad_symbol): ['RAB28', 'TRIM27', 'KRT40']

db_symbol: ACTB
Interacting partners (ad_symbol): ['ACTB', 'CFL2', 'ACTG1']

db_symbol: ACY1
Interacting partners (ad_symbol): ['ACY1']

db_symbol: ADIPOQ
Interacting partners (ad_symbol): ['PVR', 'TMEM237', 'BIK', 'BCL2L13', 'CLDN9', 'NEMP1', 'GPR42', 'CREB3L1', 'FNDC9', 'SLC35E3', 'AMIGO1', 'SLC35C2', 'CPLX4', 'FCGR1A', 'MFF', 'GPR152', 'MRM1', 'SYNE4', 'FAM209A', 'BTNL9']

db_symbol: AGXT
Interacting partners (ad_symbol): ['RFX6', 'FOSB', 'KRT31', 'CYSRT1', 'NOTCH2NL', 'NBPF19', 'KRTAP1-1', 'KRT40', 'KRT34']

db_symbol: AHCY
Interacting partners (ad_symbol): ['C1orf50']

db_symbol: AIPL1
Interacting partners (ad_symbol): ['PLSCR3']

db_symbol: ALAS2
Interacting partners (ad_symbol): ['BANP']

db_symbol: ALDOA
Interacting partners (ad_symbol): ['ALDOA']

db_symbol: AL

In [15]:
#Full snippet to save PPIs involving 247 unique genes (db_symbol): 

#Group by db_symbol and collect the unique ad_symbol values
partners_per_gene = scores_db_ad.groupby("db_symbol")["ad_symbol"].unique()

#Convert the Series into a DataFrame and rename the columns
df_partners = partners_per_gene.reset_index()
df_partners.columns = ["Gene", "Interactors"]

#Convert the array of interactors into a comma-separated string (without quotes)
df_partners["Interactors"] = df_partners["Interactors"].apply(lambda x: ", ".join(x))
df_partners.head()

Gene                                        Interactors
0  AC012254.2  GPR42, TMEM237, PIGP, TMEM86B, HERPUD2, TMEM80...
1       ACSF3                               RAB28, TRIM27, KRT40
2        ACTB                                  ACTB, CFL2, ACTG1
3        ACY1                                               ACY1
4      ADIPOQ  PVR, TMEM237, BIK, BCL2L13, CLDN9, NEMP1, GPR4...

In [16]:
#Save the result to an Excel file to work with CZ single cell data:
df_partners.to_excel("./results_11032025/Jess_PPI_21032025/PPI_1percent_Gene_partners_Final_13042025.xlsx", index=False)

In [17]:
#flat, comma-separated list for CZ Cell*Gene gene expression: 

#Group by db_symbol and collect unique interactors
partners_per_gene = scores_db_ad.groupby("db_symbol")["ad_symbol"].unique()

#Build the final list: Gene, Interactor1, Interactor2, ...
flat_list = []

for gene, interactors in partners_per_gene.items():
    flat_list.append(gene)
    flat_list.extend(interactors)

#Convert the flat list to a comma-separated string
ppi_query_list = ", ".join(flat_list)

#Optional: print first few to verify
print(ppi_query_list[:500])  


AC012254.2, GPR42, TMEM237, PIGP, TMEM86B, HERPUD2, TMEM80, UBQLN1, TM4SF18, ACSF3, RAB28, TRIM27, KRT40, ACTB, ACTB, CFL2, ACTG1, ACY1, ACY1, ADIPOQ, PVR, TMEM237, BIK, BCL2L13, CLDN9, NEMP1, GPR42, CREB3L1, FNDC9, SLC35E3, AMIGO1, SLC35C2, CPLX4, FCGR1A, MFF, GPR152, MRM1, SYNE4, FAM209A, BTNL9, AGXT, RFX6, FOSB, KRT31, CYSRT1, NOTCH2NL, NBPF19, KRTAP1-1, KRT40, KRT34, AHCY, C1orf50, AIPL1, PLSCR3, ALAS2, BANP, ALDOA, ALDOA, ALOX5, MAD1L1, NUTM1, CEP63, AMPD2, AMPD2, AMPD1, ANKRD1, ZNF446, NAG


In [18]:
len(ppi_query_list)

11399

Let's split PPI query list alphabetically by db_symbol (starting letter A–Z)— especially for querying on CellxGene & not to compromise the performance by quering 11399 genes at one go. 

We’ll:

- Group by db_symbol and get interactors.
- Create a dictionary {A: [Gene1, Int1, Int2, ..., Gene2, ...], B: [...], ..., Z: [...]}.
- Save each list into a separate A_PPI_list.txt, B_PPI_list.txt, etc.
- Print how many genes were used in each case (i.e., how many db_symbols per letter).

In [19]:
import os
from collections import defaultdict

In [20]:
#Group and collect interactors
partners_per_gene = scores_db_ad.groupby("db_symbol")["ad_symbol"].unique()

In [21]:
#Prepare output directory
os.makedirs("ppi_lists_by_letter", exist_ok=True)

#Dictionary to store flat lists A–Z
alpha_dict = defaultdict(list)
gene_count_by_letter = defaultdict(int)

#Build flat list per starting letter
for gene, interactors in partners_per_gene.items():
    first_letter = gene[0].upper()
    alpha_dict[first_letter].append(gene)
    alpha_dict[first_letter].extend(interactors)
    gene_count_by_letter[first_letter] += 1

#Save to separate files
for letter, genes_and_interactors in alpha_dict.items():
    flat_str = ", ".join(genes_and_interactors)
    filename = f"ppi_lists_by_letter/{letter}_PPI_list.txt"
    with open(filename, "w") as f:
        f.write(flat_str)
    print(f"Saved {letter}_PPI_list.txt with {gene_count_by_letter[letter]} genes")

# Optional: print a summary of total gene count per letter
print("\n Gene count summary per letter:")
for letter in sorted(gene_count_by_letter.keys()):
    print(f"{letter}: {gene_count_by_letter[letter]} genes")


Saved A_PPI_list.txt with 19 genes
Saved B_PPI_list.txt with 5 genes
Saved C_PPI_list.txt with 35 genes
Saved D_PPI_list.txt with 7 genes
Saved E_PPI_list.txt with 8 genes
Saved F_PPI_list.txt with 14 genes
Saved G_PPI_list.txt with 18 genes
Saved H_PPI_list.txt with 5 genes
Saved I_PPI_list.txt with 10 genes
Saved K_PPI_list.txt with 10 genes
Saved L_PPI_list.txt with 4 genes
Saved M_PPI_list.txt with 10 genes
Saved N_PPI_list.txt with 13 genes
Saved O_PPI_list.txt with 3 genes
Saved P_PPI_list.txt with 25 genes
Saved R_PPI_list.txt with 6 genes
Saved S_PPI_list.txt with 24 genes
Saved T_PPI_list.txt with 17 genes
Saved U_PPI_list.txt with 3 genes
Saved V_PPI_list.txt with 3 genes
Saved W_PPI_list.txt with 1 genes
Saved X_PPI_list.txt with 2 genes
Saved Z_PPI_list.txt with 5 genes

 Gene count summary per letter:
A: 19 genes
B: 5 genes
C: 35 genes
D: 7 genes
E: 8 genes
F: 14 genes
G: 18 genes
H: 5 genes
I: 10 genes
K: 10 genes
L: 4 genes
M: 10 genes
N: 13 genes
O: 3 genes
P: 25 genes


In [22]:
import os

ppi_dir = "./results_11032025/Jess_PPI_21032025/ppi_lists_by_letter"

print("Total entries (genes + interactors) in each PPI list:\n")

#Loop through all txt files in the directory
for file in sorted(os.listdir(ppi_dir)):
    if file.endswith("_PPI_list.txt"):
        file_path = os.path.join(ppi_dir, file)
        
        #Read content
        with open(file_path, "r") as f:
            content = f.read()
        
        #Split by comma, strip whitespace
        genes_and_interactors = [entry.strip() for entry in content.split(",") if entry.strip()]
        
        #Optionally use set() if you want to count unique entries only
        total_count = len(genes_and_interactors)
        
        print(f"{file}: {total_count} entries")


Total entries (genes + interactors) in each PPI list:

A_PPI_list.txt: 101 entries
B_PPI_list.txt: 66 entries
C_PPI_list.txt: 168 entries
D_PPI_list.txt: 32 entries
E_PPI_list.txt: 128 entries
F_PPI_list.txt: 106 entries
G_PPI_list.txt: 93 entries
H_PPI_list.txt: 13 entries
I_PPI_list.txt: 38 entries
K_PPI_list.txt: 73 entries
L_PPI_list.txt: 34 entries
M_PPI_list.txt: 63 entries
N_PPI_list.txt: 49 entries
O_PPI_list.txt: 11 entries
P_PPI_list.txt: 274 entries
R_PPI_list.txt: 32 entries
S_PPI_list.txt: 111 entries
T_PPI_list.txt: 91 entries
U_PPI_list.txt: 43 entries
V_PPI_list.txt: 7 entries
W_PPI_list.txt: 3 entries
X_PPI_list.txt: 6 entries
Z_PPI_list.txt: 15 entries
